

---
**Parcial Aplicaciones de Data-Science**

**Semestre 2024-01**

**Nombre del Alumno:**  Luis Felipe Poma Astete

**Código:**  202110902


---



## **CASO DE ANALISIS:** CLASIFICACIÓN DE POEMAS

El dataset consta de un conjunto de 786 poemas en ingles clasificados como Affection,Environment,Music and Death para entrenamiento y 150 para prueba.

Se pide hacer el prepocesamiento de datos, creación de la bolsa de palabras y platear e intrepretar un clasificador naive bayes con los datos.



## **Normalizacion del Texto**

**1) Creacion de Corpus**

A partir del archivos Poem_classification - test_data.csv y Poem_classification - train_data.csv, crear dos (2) corpus que consideren lo siguiente:

**Corpus #1 Train:**  Conteniendo solo el texto de entrenamiento sin etiquetas

**Corpus #2 Test:**  Conteniendo solo el texto de test sin etiquetas



**2) Normalizar corpus**

* Tokenizar y convertir a minusculas
* Eliminar los tokens que sean signos de puntuación
* Transformar cada token a su lema
* Elimine los tokens no alfabeticos y menores a tres (3) dígitos

## **BOLSA DE PALABRAS**

**3) Bag of Words**

* Devuelva un arreglo con las frecuencia de cada uno de los tokens unicos por documento. Utilice la estrategia de Vector de Frecuencias.
* Devuelva un arreglo TF-IDF para cada documento.


## **CLASIFICADOR BAYESIANO**
**4) Clasificador bayeciano**

* Realice el clasificador Naive Bayes con los vectores de frecuencia
* Realice el clasificador Naive Bayes con los vectores TF-IDF

**5) Prueba**

* Con el corpus de test pruebe el rendimiento de ambos modelos utilize las metricas Presicion, Recall y F1-SCORE

## **INTERPRETACIÓN**

**6) Interpretación de resultados**

* Explique los resultados obtenidos en el apartado 4, describa el mejor modelo e interprete el porque de su rendimiento correpondiente en relación a las metricas.











## **SOLUCIÓN**

## **Normalización del Texto**

In [1]:
'''
1) Creacion de Corpus**

A partir del archivos Poem_classification - test_data.csv y Poem_classification - train_data.csv, crear dos (2) corpus que consideren lo siguiente:

**Corpus #1 Train:**  Conteniendo solo el texto de entrenamiento sin etiquetas

**Corpus #2 Test:**  Conteniendo solo el texto de test sin etiquetas

2) Normalizar corpus


* Tokenizar y convertir a minusculas
* Eliminar los tokens que sean signos de puntuación
* Transformar cada token a su lema
* Elimine los tokens no alfabeticos y menores a tres (3) dígitos
'''

'\n1) Creacion de Corpus**\n\nA partir del archivos Poem_classification - test_data.csv y Poem_classification - train_data.csv, crear dos (2) corpus que consideren lo siguiente:\n\n**Corpus #1 Train:**  Conteniendo solo el texto de entrenamiento sin etiquetas\n\n**Corpus #2 Test:**  Conteniendo solo el texto de test sin etiquetas\n\n2) Normalizar corpus\n\n\n* Tokenizar y convertir a minusculas\n* Eliminar los tokens que sean signos de puntuación\n* Transformar cada token a su lema\n* Elimine los tokens no alfabeticos y menores a tres (3) dígitos\n'

Primero cargamos los datos

In [2]:
import pandas as pd

In [3]:
data_train = pd.read_csv("./Poem Classification/Poem_classification - train_data.csv")
data_test = pd.read_csv("./Poem Classification/Poem_classification - test_data.csv")

In [4]:
data_train.head(2)

Genre                                               Poem
0  Music                                                NaN
1  Music                In the thick brushthey spend the...

In [5]:
data_test.head(2)

Genre                                               Poem
0  Music  A woman walks by the bench I’m sitting onwith ...
1  Music  Because I am a boy, the untouchability of beau...

Limpiamos los datos nulos de `train`

In [6]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 841 entries, 0 to 840
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Genre   841 non-null    object
 1   Poem    837 non-null    object
dtypes: object(2)
memory usage: 13.3+ KB


In [7]:
data_train.dropna(inplace=True)

In [8]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 837 entries, 1 to 840
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Genre   837 non-null    object
 1   Poem    837 non-null    object
dtypes: object(2)
memory usage: 19.6+ KB


Ahora normalizamos el texto

In [9]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')


lemmatizer = WordNetLemmatizer()
stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\pms_l\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\pms_l\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\pms_l\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [10]:
def normalize_text(text):
    # Tokenizar y convertir a minúsculas
    tokens = word_tokenize(text.lower())
    
    # Eliminar signos de puntuación y tokens no alfabéticos
    tokens = [token for token in tokens if token.isalpha()]

    # Lematizar y eliminar palabras de menos de tres caracteres
    tokens = [lemmatizer.lemmatize(token) for token in tokens if len(token) > 3]

    # Eliminar stopwords
    tokens = [token for token in tokens if token not in stop_words]

    return ' '.join(tokens)

In [11]:
data_train['normalized_text'] = data_train['Poem'].apply(normalize_text)
data_test['normalized_text'] = data_test['Poem'].apply(normalize_text)

In [12]:
# etiquetas de entrenamiento y prueba
label_train = data_train["Genre"]
label_test = data_test["Genre"]

# texto de entrenamiento sin etiquetas
corpus_train = data_train["normalized_text"].tolist()

# texto de prueba sin etiquetas
corpus_test = data_test["normalized_text"].tolist()

# Ejemplo de visualización de los primeros documentos normalizados en el corpus de entrenamiento
print("Ejemplo de documentos normalizados en el corpus de entrenamiento:")
for i in range(5):
    print(f"Documento {i+1}: {corpus_train[i]}")

Ejemplo de documentos normalizados en el corpus de entrenamiento:
Documento 1: thick brushthey spend hottest part soaking hoovesin trickle mountain water ravine hoardson behalf oleander
Documento 2: storm generous something easy surrender sitting window step garden bored
Documento 3: mendieta carry around matin star hold hand would wake radiate shimmer gleam morning would measure wingspan idea taking night would
Documento 4: sherrard portent memory stevenshow hard carry score adult back look carrion need distressof loyalty requires pain
Documento 5: marley bavaria november brilliant morning fishing boat dream dying midwinter world covered light dream


## **BOLSA DE PALABRAS**

In [13]:
'''
3) Bag of Words

* Devuelva un arreglo con las frecuencia de cada uno de los tokens unicos por documento. Utilice la estrategia de Vector de Frecuencias.
* Devuelva un arreglo TF-IDF para cada documento.
'''

'\n3) Bag of Words\n\n* Devuelva un arreglo con las frecuencia de cada uno de los tokens unicos por documento. Utilice la estrategia de Vector de Frecuencias.\n* Devuelva un arreglo TF-IDF para cada documento.\n'

In [14]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

Hallamos el documento con mayor tamaño, luego tomarlo como parametro de cuantas palabras como maximo tendra nuestro vocabulario, asi evitando que existan muchos ceros en nuestro `BOW`.

In [15]:
# Encontramos el maximo de tokens unicos en el corpus de entrenamiento
max_len = max([len(doc.split()) for doc in corpus_train])

In [16]:
count_vectorizer = CountVectorizer(max_features=max_len + 10, stop_words="english")
tfidf_vectorizer = TfidfVectorizer(max_features=max_len + 10, stop_words="english")

Ahora creamos nuestro ``Bag Of Words``

In [17]:
# Ajustar CountVectorizer al corpus de entrenamiento
X_train_count = count_vectorizer.fit_transform(corpus_train).toarray()
# Ajustar TfidfVectorizer al corpus de entrenamiento
X_train_tfidf = tfidf_vectorizer.fit_transform(corpus_train).toarray()

# Ejemplo de visualización del vocabulario generado por CountVectorizer
bow_vectorizer = count_vectorizer.get_feature_names_out()
print(f"Ejemplo de tokens únicos (vocabulario) generado por CountVectorizer:")
print(bow_vectorizer[:50], "\n")  # Mostrar los primeros 50 tokens únicos

print(corpus_train[0])
# Ejemplo de visualización de la matriz de frecuencias (Bag of Words) para el primer documento de entrenamiento
print(
    "\nEjemplo de matriz de frecuencias (Bag of Words) para el primer documento de entrenamiento:"
)
# imprimir en forma de matriz
X_train_count[0]

Ejemplo de tokens únicos (vocabulario) generado por CountVectorizer:
['away' 'black' 'blue' 'body' 'child' 'come' 'dark' 'dead' 'death' 'door'
 'earth' 'eye' 'face' 'field' 'flower' 'foot' 'grass' 'green' 'hand'
 'heart' 'hour' 'know' 'leaf' 'life' 'light' 'like' 'little' 'long' 'look'
 'love' 'make' 'moon' 'morning' 'mother' 'night' 'river' 'said' 'song'
 'star' 'summer' 'tell' 'thing' 'think' 'time' 'tree' 'water' 'white'
 'wind' 'woman' 'word'] 

thick brushthey spend hottest part soaking hoovesin trickle mountain water ravine hoardson behalf oleander

Ejemplo de matriz de frecuencias (Bag of Words) para el primer documento de entrenamiento:


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0], dtype=int64)

In [18]:
# Ejemplo de visualización de la matriz TF-IDF para el primer documento de entrenamiento
bow_tfidf = tfidf_vectorizer.get_feature_names_out()
print(f"\nEjemplo de tokens únicos (vocabulario) generado por TfidfVectorizer:")
print(bow_tfidf[:50])  # Mostrar los primeros 50 tokens únicos

# Ejemplo de visualización de la matriz TF-IDF para el primer documento de entrenamiento
print("\nEjemplo de matriz TF-IDF para el primer documento de entrenamiento:")
X_train_tfidf[0]


Ejemplo de tokens únicos (vocabulario) generado por TfidfVectorizer:
['away' 'black' 'blue' 'body' 'child' 'come' 'dark' 'dead' 'death' 'door'
 'earth' 'eye' 'face' 'field' 'flower' 'foot' 'grass' 'green' 'hand'
 'heart' 'hour' 'know' 'leaf' 'life' 'light' 'like' 'little' 'long' 'look'
 'love' 'make' 'moon' 'morning' 'mother' 'night' 'river' 'said' 'song'
 'star' 'summer' 'tell' 'thing' 'think' 'time' 'tree' 'water' 'white'
 'wind' 'woman' 'word']

Ejemplo de matriz TF-IDF para el primer documento de entrenamiento:


array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
       0.])

In [19]:
# Obtener los arreglos de frecuencias y TF-IDF para el corpus de prueba
X_test_count = count_vectorizer.transform(corpus_test)
X_test_tfidf = tfidf_vectorizer.transform(corpus_test)

## **CLASIFICADOR BAYESIANO**

In [20]:
'''
4) Clasificador bayeciano

* Realice el clasificador Naive Bayes con los vectores de frecuencia
* Realice el clasificador Naive Bayes con los vectores TF-IDF

5) Prueba

* Con el corpus de test pruebe el rendimiento de ambos modelos utilize las metricas Presicion, Recall y F1-SCORE

'''

'\n4) Clasificador bayeciano\n\n* Realice el clasificador Naive Bayes con los vectores de frecuencia\n* Realice el clasificador Naive Bayes con los vectores TF-IDF\n\n5) Prueba\n\n* Con el corpus de test pruebe el rendimiento de ambos modelos utilize las metricas Presicion, Recall y F1-SCORE\n\n'

In [21]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report

In [22]:
# Inicializar y entrenar el clasificador Naive Bayes con vectores de frecuencia
nb_freq = MultinomialNB()

nb_freq.fit(X_train_count, label_train)

# Inicializar y entrenar el clasificador Naive Bayes con vectores TF-IDF
nb_tfidf = MultinomialNB()
nb_tfidf.fit(X_train_tfidf, label_train)

# Predecir las etiquetas para el corpus de prueba usando ambos modelos
y_pred_freq = nb_freq.predict(X_test_count)
y_pred_tfidf = nb_tfidf.predict(X_test_tfidf)

In [23]:
# Evaluar el rendimiento de los modelos con métricas de clasificación
print("Resultados del clasificador Naive Bayes con vectores de frecuencia:")
print(classification_report(label_test, y_pred_freq))

Resultados del clasificador Naive Bayes con vectores de frecuencia:
              precision    recall  f1-score   support

   Affection       0.94      0.34      0.50       100
       Death       0.23      0.38      0.29        13
 Environment       0.30      0.44      0.35        25
       Music       0.13      0.58      0.21        12

    accuracy                           0.38       150
   macro avg       0.40      0.44      0.34       150
weighted avg       0.71      0.38      0.43       150



In [24]:
print("\nResultados del clasificador Naive Bayes con vectores TF-IDF:")
print(classification_report(label_test, y_pred_tfidf))


Resultados del clasificador Naive Bayes con vectores TF-IDF:
              precision    recall  f1-score   support

   Affection       0.90      0.19      0.31       100
       Death       0.19      0.38      0.25        13
 Environment       0.24      0.36      0.29        25
       Music       0.11      0.58      0.18        12

    accuracy                           0.27       150
   macro avg       0.36      0.38      0.26       150
weighted avg       0.67      0.27      0.29       150



## **INTERPRETACIÓN**

In [25]:
'''6) Interpretación de resultados**

* Explique los resultados obtenidos en el apartado 4, describa el mejor modelo e interprete el porque de su rendimiento correpondiente en relación a las metricas.
'''

'6) Interpretación de resultados**\n\n* Explique los resultados obtenidos en el apartado 4, describa el mejor modelo e interprete el porque de su rendimiento correpondiente en relación a las metricas.\n'

**Mejor Modelo y Interpretación:**

- Al comparar ambos modelos, el rendimiento del modelo con vectores de frecuencia parece ser ligeramente mejor en términos de precisión, recall y F1-score en la mayoría de las categorías en comparación con el modelo TF-IDF. Sin embargo, ambos modelos muestran limitaciones significativas, como una baja capacidad para identificar correctamente poemas en algunas categorías y una alta tasa de falsos positivos. El modelo con vectores de frecuencia tiende a generalizar un poco mejor en la mayoría de las categorías.
  
- El rendimiento deficiente en ambos modelos puede atribuirse a varios factores, como la complejidad y la naturaleza subjetiva de la clasificación de poemas en estas categorías específicas, así como la falta de datos de entrenamiento representativos (con solo 786 registros) y la simplicidad inherente del modelo Naive Bayes, que tiene sus limitaciones. En este caso, sería recomendable el uso de modelos más avanzados basados en redes neuronales para mejorar el rendimiento.
  
- Además, en intentos anteriores no se utilizó el parámetro `max_features` en los Bag of Words (BOWs), lo que resultó en bolsas de palabras con muchos ceros. Como consecuencia, al entrenar los modelos, se obtuvieron métricas muy dispares entre las categorías y una precisión (accuracy) por debajo de `0.2` con ambos BOWs. Sin embargo, al definir `max_features` como un parámetro en CountVectorizer, mejoró el rendimiento de los modelos. El modelo que utilizaba `CountVectorizer` aumentó su precisión a `0.39`, mientras que el modelo con `TfidfVectorizer` mejoró su precisión a `0.26`.

Una vez desarrolladas todas las operaciones solicitadas, no olvide adjuntar como respuesta a esta evaluación:

* **Este Notebook** con sus datos personales conteniendo su solución
